### PR TITLE
feat: support retry-after header

### PR DESCRIPTION
**Pull request by blakeyp**
Adds functionality to check for Retry-After header in error response and parse the value if it's available (as per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After)

- Handles number and Date header values
- Delays by parsed value up to max number of retries
- Rejects immediately if header value could not be parsed/is invalid
- Rejects immediately if header value is greater than configured `maxRetryAfter` (default 5 mins)
- Feature can be toggled off via `checkRetryAfter` config param

### CODE REVIEW

File: test/index.ts [hunk 1]
```diff
@@ -396,4 +398,32 @@ describe('retry-axios', () => {
...
   });
+
+  it('should parse Retry-After header for retry delay if available', async () => {
+    const scopes = [
+      nock(url).get('/').reply(429, undefined, {'Retry-After': '1'}),
+      nock(url).get('/').reply(200, 'toast'),
+    ];
+    interceptorId = rax.attach();
+    const res = await axios({url});
+    assert.strictEqual(res.data, 'toast');
```
Comment (JustinBeckwith) [inline test/index.ts:409]: This isn't actually testing if the delays work.  I know it's annoying, but could I trouble you to set up a test with a sinon timer, and verify it's doing what we expect?  I've had to write similar tests in the past if you're looking for an example:
https://github.com/JustinBeckwith/linkinator/blob/main/test/test.retry.ts#L83

File: test/index.ts [hunk 2]
```diff
@@ -396,4 +398,32 @@ describe('retry-axios', () => {
...
+
+  it('should not retry if Retry-After value greater than maxRetryAfter', async () => {
+    const scope = nock(url)
+      .get('/')
+      .reply(429, undefined, {'Retry-After': '2'});
+    interceptorId = rax.attach();
+    try {
+      const cfg: rax.RaxConfig = {url, raxConfig: {maxRetryAfter: 1000}};
+      await axios(cfg);
+    } catch (e) {
```
Comment (JustinBeckwith) [inline test/index.ts:421]: I know I may be using this pattern elsewhere in the test, but it's less than ideal.  Would you mind using an `assert.rejects` instead?

**Review by JustinBeckwith — COMMENTED** [review]
This is really great - thank you!  Mostly a few test nits that would be nice to clean up before landing.

File: src/index.ts [hunk 3]
```diff
@@ -124,6 +134,26 @@ function normalizeArray<T>(obj?: T[]): T[] | undefined {
...
+function parseRetryAfter(header: string): number | undefined {
+  // Header value may be string containing integer seconds
+  const value = parseInt(header);
+  if (!Number.isNaN(value)) {
+    return value * 1000;
+  }
+  // Or HTTP date time string
+  const dateTime = Date.parse(header);
+  if (!Number.isNaN(dateTime)) {
+    return dateTime - Date.now() + 3000; // Add 3sec buffer
```
[the remark this answers is not in the document]
Comment (blakeyp) [reply src/index.ts]: Removed that now as yeah it's not necessary :)

File: test/index.ts (hunk 1 above)
Comment (blakeyp) [reply test/index.ts:429]: Done using the same pattern - also set a higher default retry delay on the rax config to make sure the Retry-After value is actually being used

File: test/index.ts (hunk 2 above)
Comment (blakeyp) [reply test/index.ts]: Done 👍

File: test/index.ts [hunk 4]
```diff
@@ -78,7 +82,8 @@ describe('retry-axios', () => {
     assert.fail('Expected to throw');
   });
 
-  it('should retry at least the configured number of times', async () => {
+  it('should retry at least the configured number of times', async function () {
+    this.timeout(10000);
```
Comment (blakeyp) [inline test/index.ts:86]: Had to change the timeouts to the old function style after getting a weird error on the ci tests - https://github.com/JustinBeckwith/retry-axios/pull/142/checks?sha=f6f6c8d1bbe8b2a6da5301ee457fd6fe0970e121

**Comment by simllll** [comment]
Anything we can do to get this feature merged? :)

**Comment by akshayrohatgi-hotstar** [comment]
@JustinBeckwith @blakeyp Any updates on this PR?

**Review by JustinBeckwith — APPROVED** [review]

**Comment by akshayrohatgi-hotstar** [comment]
@JustinBeckwith @blakeyp Thanks a lot :)

**Pull request merged**